### DISCUSSION TITLE
`vcpkg env` doesn't preserve the VCPKG_VISUAL_STUDIO_PATH environment variable

### DISCUSSION THREAD

**Discussion by metathinker** [microsoft/vcpkg#19247] [Ideas]
### To see this problem in action:

1. On a Windows computer with Visual Studio 2017, set up VCpkg following the [Quick Start](https://github.com/microsoft/vcpkg/blob/2b78cac373da1bde47d4c99d3c7c5f96398181a4/README.md#quick-start) instructions.
2. In a new cmd.exe command prompt, enter:
   `set VCPKG_VISUAL_STUDIO_PATH=C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise`
   (change "Enterprise" to whatever your VS edition is)
3. Then enter:
   [`...\vcpkg.exe env`](https://github.com/microsoft/vcpkg/blob/2b78cac373da1bde47d4c99d3c7c5f96398181a4/toolsrc/src/vcpkg/help.cpp#L108-L109)

Currently, the new environment clears the `VCPKG_VISUAL_STUDIO_PATH` environment variable, which you can confirm by running `set VCPKG_VISUAL_STUDIO_PATH`. This means that if you then run `...\vcpkg.exe install some-package-name`, VCpkg is not guaranteed to use the version of Visual Studio you specified.

This is especially confusing since `vcpkg env` has set you up so that if you directly run the Microsoft Visual C++ compiler with `cl.exe`, you _will_ get MSVC v141 taken from the VS 2017 installation path. But if you run `vcpkg install whatever`, the package could still be built with a different MSVC toolchain!

### Proposed solution:

In an ideal world, I'd like `vcpkg env` to preserve the value of `VCPKG_VISUAL_STUDIO_PATH` set in `vcpkg.exe`'s original environment.

### Describe alternatives considered:

- Obviously, if you don't run `vcpkg env` then you won't have this problem. :)
  However, you'll lose the benefits of a cleaned-up PATH and so on.
- After each time you run `vcpkg env`, you could remember to `set VCPKG_VISUAL_STUDIO_PATH=[whatever]` again. But that's something you'd have to remember - and to figure out the first time!
- I was actually thinking of submitting a PR just to document the current behavior above in the [documentation page for `VCPKG_VISUAL_STUDIO_PATH`](https://github.com/microsoft/vcpkg/blob/2b78cac373da1bde47d4c99d3c7c5f96398181a4/docs/users/config-environment.md#vcpkg_visual_studio_path). That way, people like me who are surprised by the current behavior will be warned, while people who prefer or rely on the current behavior not changing will remain happy.

### Additional context:

I ran into this problem while trying to work around the fact that Microsoft C++ REST SDK does not currently build for the Universal Windows Platform (`[cpu]-uwp` triples) with Visual Studio 2019's MSVC v142. See #7540 ; https://github.com/microsoft/cpprestsdk/issues/1227 . It was not fun to debug.

**Comment by MVoz** [comment]
Variables need to be prescribed in triplet

All external will be ignored

**Comment by metathinker** [comment]
> Variables need to be prescribed in triplet
>
>All external will be ignored

I'm not sure what you mean? I think I misunderstand you.

Yes, it's true that the VCPKG_VISUAL_STUDIO_PATH environment variable will be overridden if the same is set in your triplet's cmake file. But you _can_ set both. It is not that VCPKG_VISUAL_STUDIO_PATH is totally ignored as an env var.

If VCPKG_VISUAL_STUDIO_PATH was only a triplet variable, there would be no point for this request since I would just write my own triplet file as that would be the only way to change compiler version. But it can be both in environment or in triplet.cmake, so I think the request at least makes sense.

Or maybe you intend to suggest I should write my own triplet?

**Comment by MVoz** [comment]
vcpkg/triplet/x64-windows.cmake <<-- That's the triplet. Prescribe variables there

Read the documentation if you're really interested in understanding

**Comment by metathinker** [comment]
I do understand that is where the triplet config file is, and that if you set the VCPKG_VISUAL_STUDIO_PATH there it overrides the env var of the same name. I don't see why it's relevant to my query which is about the env var. Given we can use both, I would like to have the env var work better. 

I think I tried to explain this already but I was not clear.

**Comment by MVoz** [comment]
https://github.com/microsoft/vcpkg/blob/master/docs/users/triplets.md#vcpkg_visual_studio_path

1.
`set(VCPKG_VISUAL_STUDIO_PATH "C:\\Program Files (x86)\\Microsoft Visual Studio\\Preview\\Community")`

2.
The Visual Studio 2017 platform toolset is v141.

`set(VCPKG_PLATFORM_TOOLSET v141)`

3.
https://github.com/microsoft/vcpkg/blob/300e21d59ebfe42e118cf4e97887f0680fbcfa2f/toolsrc/src/vcpkg.cpp#L126

ENV SYSTEM, not USER ENV

need more?
